src/schema.ts

```diff
@@ -0,0 +1,379 @@
+	 * This doesn't have to be the same as the author of the program.
+	 */
+	author: string;
+}
+
+/**
+ * An untimed event that changes the state of one channel
+ * or manually drops a marble.
+ *
+ * Events only change one thing. Multiple simultaneous changes must
+ * be represented byt multiple events with the same tick.
+ */
+export type Event =
+	| DropEvent
+	| MachineEvent
+	| VibraphoneEvent
+	| HihatMachineEvent
+	| HihatEvent
+	| BassEvent;
+
+/** Represents an event occurring in time (non-tempo dependent). */
+export type TimedEvent = BaseTimedEvent & Event;
+/** Represents information associated with all timed events. */
+export interface BaseTimedEvent {
+	/** The time (in seconds) that this event occurs. */
+	time: number;
+}
+
+/**
+ * Represents the overall state of the machine. Everything that's
+ * not specific to a particular instrument is represented here.
+ */
+export interface MachineState {
+	/** Whether or not particular instrument from [[Channels]] is muted. */
+	mute: { [C in Channel]?: boolean };
+	/** The beats per minute the machine is operating at. */
+	bpm: number;
+	/** Whether or not the programming wheel is spinning. */
+	flywheelConnected: boolean;
+}
+
+/**
+ * All possible events relating to the machine itself.
+ */
+export type MachineEvent = MachineMuteEvent;
+/** An event representing the muting or unmuting of a [[Channel]] */
+export interface MachineMuteEvent {
+	kind: "machine_mute";
+	/** The channel affected. */
+	channel: Channel;
+	/** Whether or not the channel should be muted or unmuted. */
+	muted: boolean;
+}
+/**
+ * An event representing a change in the machine's BPM.
+ *
+ * This change will be instantaneous, so if a gradual change is
+ * desired, it must be achieved by specifying small increments of
+ * change to the BPM. This may change in the future.
+ */
+export interface MachineTempoEvent {
+	kind: "machine_tempo";
+	/** The BPM to set the machine to. */
+	bpm: number;
+}
+/**
+ * An event representing the flywheel getting connected or disconnected.
+ *
+ * This equates the the programming wheel starting or stopping.
+ */
+export interface FlywheelConnectedEvent {
+	kind: "machine_flywheelConnected";
+	/**
+	 * Whether or not the flywheel is connected and thus also
+	 * whether or not the programming wheel is turning.
+	 */
+	connected: boolean;
+}
+
+/**
+ * Represents every channel that can be played on the vibraphone.
+ *
+ * We decided to start at one since the [[BassString]] type also
+ * starts at one. We also decided to only represent playable
+ * channels and not actual channels (two per note) because that is
+ * an implementation detail of the physical MMX.
+ */
+export type VibraphoneChannel = 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10 | 11;
+
+/**
+ * Represents the state pertaining to the vibraphone.
+ */
+export interface VibraphoneState {
+	/** Whether or not the vibrato is engaged */
+	vibratoEnabled: boolean;
+	/** Which vibrato gear is selected. */
+	vibratoSpeed: number;
```

Comment:
This is supposed to be a multiplier of some sort.
Currently, no fancy gear-to-speed conversion is planned. This may change when it's implemented, however.

For now, it should just be a 0-1 multiplier of vibrato speed.

Reply:
Okay. I think if that's the case, we shouldn't have a separate `vibratoEnabled` property when you could simply set `vibratoSpeed` to 0. But this is conversation we can carry on somewhere else.

Reply:
Me too; there was some back and forth on this and we settled here. This can be dealt with by whomever implements vibrato 🤷‍♂️.